### DISCUSSION TITLE
Request timeouts?

### DISCUSSION THREAD

**Discussion by MSP-Greg** [puma/puma#2987] [Announcements]
From a recent announcement about [pitchfork](https://github.com/Shopify/pitchfork):

"For instance you don't want to refork when another thread is currently processing a request, you'd risk leaving some global resource in a corrupted state. So you'd first need to stop accepting traffic and wait for all requests to complete.  The problem is, Puma doesn't have a request timeout, so if the request never complete, what do you do?"

I need to check, but I thought Puma will force connections closed on restarts, but that may not happen in Worker 0 when refork occurs.  Not.sure.

Also, not that familiar with 'Copy-on-Write' particulars, but the workers are reinitialized, so I don't know about "you'd risk leaving some global resource in a corrupted state".

Regardless, back to request timeouts.  At present, Puma has `Puma::DSL#first_data_timeout`, which is a good timeout to have, but I wonder if there should be more?  Some thoughts:

* 'Request/Header lines' timeout
* 'Request/Header lines' first data timeout
* Body timeout, only used if the request has a body
* Body first data timeout, only used if the request has a body
* Total request timeout

Some Puma refs:

PR #1111 'Config first data timeout'

Re pitchfork, see also https://www.reddit.com/r/ruby/comments/xwcvty/a_new_web_server_github_shopifypitchfork/ and https://nitter.net/_byroot/status/1577647475686514689.  Also, the Slack 'How to Open Source' 'language-ruby' channel.

**Comment by nateberkopec** [comment]
> Total request timeout

This is always going to be difficult and there's a reason it was never included in Puma to begin with. App threads can be doing basically anything, checking out any number of resources/locks, which you nuke/corrupt when you destroy that thread unexpectedly.

@matthewd showed me some code a while ago that basically implemented a request timeout like this:

1. Timeout triggers for a thread. Thread is killed.
2. Puma stops accepting new requests. Existing requests allowed to complete.
3. Process is restarted.

This increases the safety of timing out a thread. Basically every big company I know of does this when timing out Unicorns (process recycling on timeout). It's a bit dangerous though, because people usually have really poor instrumentation on timeouts and so they end up with a process restarting every 10 seconds but they have no idea it's happening.

the first 4 proposed timeouts are much less controversial and we can proceed on them.

**Reply by MSP-Greg** [reply]
First of all, re pitchfork and some of the comments re Puma, see [Fork-Worker Cluster Mode](https://msp-greg.github.io/puma/file.fork_worker.html).  I guess I find some of the comments re Puma to be somewhat vague.

Anway, to clarify, when I mentioned timeouts, it was from the perspective of dividing processing into three sections:
1. Acquire the request.
2. Pass the request to the app, wait for return info.
3. Send the response.

Timeouts are only applicable to step 1.  Adding timeouts to step 2 is messy, as you've mentioned.  Not sure if it's a Puma issue, but maybe a hook to pass the thread when a timeout is exceeded, assuming the hook will clean up the app as needed.  Not.sure.

Re the timeouts listed above, rather than using a 'first data' timeout, maybe an 'interval timeout'?  So, there would be interval and total timeouts for both 'header lines' and 'body'.  If any of timers are exceeded, return a 408.  That's assuming it can be done cleanly.

I've spent a lot of time with step 3, I'm doing a step 1 refresher...

**Reply by nateberkopec** [reply]
Sounds good. I'm OK with more clarity and timeouts in Step 1 and 3.
